Exception in deferred callback causes Node process crash

Dear maintainers,

The unprotected use of `setTimeout` quoted below causes hard Node process crashes when the deferred function throws an exception.

https://github.com/braintree/wrap-promise/blob/952eecaa3f625dcc1687dcf0a897dd718fb3d0df/lib/deferred.js#L8-L10

This happened in one of our Node-based alerting systems, with the log shown below:

    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: /opt/mycompany/sswallboard/node_modules/braintree/lib/braintree/transaction.js:117
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: this.creditCard = new CreditCard(attributes.creditCard);
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: ^
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: TypeError: Cannot read property 'creditCard' of undefined
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: at new Transaction (/opt/mycompany/sswallboard/node_modules/braintree/lib/braintree/transaction.js:117:49)
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: at /opt/mycompany/sswallboard/node_modules/braintree/lib/braintree/gateway.js:108:26
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: at Timeout._onTimeout (/opt/mycompany/sswallboard/node_modules/@braintree/wrap-promise/lib/deferred.js:9:10)
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: at ontimeout (timers.js:498:11)
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: at tryOnTimeout (timers.js:323:5)
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: at Timer.listOnTimeout (timers.js:290:5)
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: npm ERR! code ELIFECYCLE
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: npm ERR! errno 1
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: npm ERR! Wallboard@2.59.0 start: `node start.js`
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: npm ERR! Exit status 1
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: npm ERR!
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: npm ERR! Failed at the Wallboard@2.59.0 start script.
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: npm ERR! A complete log of this run can be found in:
    Mar 18 17:34:44 wallboard.internal.mycompany.com wallboard.sh[21603]: npm ERR!     /var/opt/mycompany/sswallboard/.npm/_logs/2019-03-18T16_34_44_525Z-debug.log
    Mar 18 17:34:44 wallboard.internal.mycompany.com systemd[1]: sswallboard.service: main process exited, code=exited, status=1/FAILURE

Currently, there is nothing the client code can do to avoid crashing on error, except top-level hacks like the [`uncaughtException` event handler](https://nodejs.org/api/process.html#process_event_uncaughtexception).

My first suggestion to circumvent this issue would be to add an optional `errorHandler` parameter to `deferred`. E.g, if the last argument is is a function, use it as an error handler.

However, since `deferred` is already used in `wrapPromise` to delay a callback function, I am not sure whether the error handler in this case should be the callback itself, or something entirely different - e.g `console.err`.

I'm looking forward to reading what is your take on this issue.
Kind regards - and thank you for your work on the Braintree SDK!

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+# unreleased
+*Breaking Changes*
+
+* Errors that are thrown in callback functions are now logged to the console (#4)
+
 # 1.1.1
 
 * Publish to npm as scoped package

--- a/lib/deferred.js
+++ b/lib/deferred.js
@@ -6,7 +6,14 @@ function deferred(fn) {
     var args = arguments;
 
     setTimeout(function () {
-      fn.apply(null, args);
+      try {
+        fn.apply(null, args);
+      } catch (err) {
+        /* eslint-disable no-console */
+        console.log('Error in callback function');
+        console.log(err);
+        /* eslint-enable no-console */
+      }
     }, 1);
   };
 }

--- a/test/deferred.js
+++ b/test/deferred.js
@@ -16,7 +16,7 @@ describe('deferred', function () {
     expect(fn).not.to.be.called;
   });
 
-  it('can pass arguments to the delayed function', function (done) {
+  it('can pass arguments to the deferred function', function (done) {
     var fn = this.sandbox.spy(function (a, b) {
       expect(arguments.length).to.equal(2);
       expect(a).to.equal(1);
@@ -29,5 +29,29 @@ describe('deferred', function () {
     def(1, 2);
 
     expect(fn).not.to.be.called;
+  });
+
+  it('sends message to console if function throws an error', function (done) {
+    var def;
+    var error = new Error('simulated error, disregard in test output');
+
+    function funcThatThrows() {
+      throw error;
+    }
+    def = deferred(funcThatThrows);
+
+    this.sandbox.spy(console, 'log');
+
+    def();
+
+    setTimeout(function () {
+      /* eslint-disable no-console */
+      expect(console.log).to.be.calledTwice;
+      expect(console.log).to.be.calledWith('Error in callback function');
+      expect(console.log).to.be.calledWith(error);
+      /* eslint-enable no-console */
+
+      done();
+    }, 5);
   });
 });

